File: 0086 - Partition List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode p1, p2, *t1 = &p1, *t2 = &p2;
        while(head){
            ListNode*& t = head->val<x?t1:t2;
            t->next = head, t = t->next;
            head = head->next;
        }
        t1->next = p2.next, t2->next = nullptr;
        return p1.next;
    }
};

File: 0086 - Partition List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        p1 = t1 = ListNode()
        p2 = t2 = ListNode()
        while head:
            if head.val < x:
                t1.next = head
                t1 = t1.next
            else:
                t2.next = head
                t2 = t2.next
            head = head.next
        t1.next, t2.next = p2.next, None
        return p1.next